File: solutions/fortran/allergies/allergies.f90

module allergies
   implicit none

   private
   character(len=12), dimension(8) :: allergens = [character(len=12) :: &
      "eggs",         & ! (1)
      "peanuts",      & ! (2)
      "shellfish",    & ! (4)
      "strawberries", & ! (8)
      "tomatoes",     & ! (16)
      "chocolate",    & ! (32)
      "pollen",       & ! (64)
      "cats"]           ! (128)

   public allergicTo, allergicList

contains
   logical function allergicTo(allergy_str, allergy_key)
      character(len=*), intent(in) :: allergy_str
      integer, intent(in) :: allergy_key

      ! determine the allergy score value for given string
      integer :: code, i
      do i = 1,size(allergens)
         if ( allergens(i) == allergy_str ) then
            code = 2**(i-1)
            exit
         end if
      end do

      allergicTo = ( iand(allergy_key, code) == code )
   end function

   function allergicList(allergy_key)
      integer, intent(in) :: allergy_key
      character(len=100) :: allergicList
      integer :: code, i

      ! check allergy score for each value and concatenate matching keys
      allergicList = ''
      do i = 1,size(allergens)
         code = 2**(i-1)
         if ( iand(allergy_key, code) == code ) then
            if ( trim(allergicList) == '' ) then
               allergicList = allergens(i)
            else
               allergicList = trim(allergicList)//' '//allergens(i)
            end if
         end if
      end do
   end function
end module

File: solutions/erlang/allergies/src/allergies.erl
-module(allergies).

-export([is_allergic_to/2, allergies/1, test_version/0]).

is_allergic_to(Allergy, Score) ->
    lists:member(Allergy, allergies(Score)).

allergies(Score) ->
    allergies(Score, []).

allergies(Score, Allergies) when (Score >= 256) ->
    allergies(Score - 256, Allergies);
allergies(Score, Allergies) when (Score >= 128) ->
    allergies(Score - 128, ['cats' | Allergies]);
allergies(Score, Allergies) when (Score >= 64) ->
    allergies(Score - 64, ['pollen' | Allergies]);
allergies(Score, Allergies) when (Score >= 32) ->
    allergies(Score - 32, ['chocolate' | Allergies]);
allergies(Score, Allergies) when (Score >= 16) ->
    allergies(Score - 16, ['tomatoes' | Allergies]);
allergies(Score, Allergies) when (Score >= 8) ->
    allergies(Score - 8, ['strawberries' | Allergies]);
allergies(Score, Allergies) when (Score >= 4) ->
    allergies(Score - 4, ['shellfish' | Allergies]);
allergies(Score, Allergies) when (Score >= 2) ->
    allergies(Score - 2, ['peanuts' | Allergies]);
allergies(Score, Allergies) when (Score >= 1) ->
    allergies(Score - 1, ['eggs' | Allergies]);
allergies(_Score, Allergies) ->
    Allergies.

test_version() -> 1.
